File: memo/src/main/java/com/zeoflow/memo/MemoBuilder.java
package com.zeoflow.memo;

import android.content.Context;

import com.zeoflow.zson.Zson;

import static com.zeoflow.memo.MemoApplication.getContext;

@SuppressWarnings({"unused", "RedundantSuppression"})
public class MemoBuilder
{

    /**
     * NEVER ever change STORAGE_TAG_DO_NOT_CHANGE and TAG_INFO.
     * It will break backward compatibility in terms of keeping previous data
     */
    private static final String STORAGE_TAG_DO_NOT_CHANGE = "MemoStorage";

    private final Context context;
    private IStorage cryptoStorage;
    private IConverter converter;
    private IParser parser;
    private Encryption encryption;
    private ISerializer serializer;
    private ILogInterceptor logInterceptor;

    public MemoBuilder()
    {
        this(getContext());
    }
    public MemoBuilder(Context context)
    {
        MemoUtils.checkNull("Context", context);

        this.context = context.getApplicationContext();
    }
    ILogInterceptor getLogInterceptor()
    {
        if (logInterceptor == null)
        {
            logInterceptor = message ->
            {
                //empty implementation
            };
        }
        return logInterceptor;
    }
    public MemoBuilder setLogInterceptor(ILogInterceptor logInterceptor)
    {
        this.logInterceptor = logInterceptor;
        return this;
    }
    IStorage getStorage()
    {
        if (cryptoStorage == null)
        {
            cryptoStorage = new SharedPreferencesStorage(context, STORAGE_TAG_DO_NOT_CHANGE);
        }
        return cryptoStorage;
    }
    public MemoBuilder setStorage(IStorage storage)
    {
        this.cryptoStorage = storage;
        return this;
    }
    IConverter getConverter()
    {
        if (converter == null)
        {
            converter = new MemoConverter(getParser());
        }
        return converter;
    }
    public MemoBuilder setConverter(IConverter converter)
    {
        this.converter = converter;
        return this;
    }
    IParser getParser()
    {
        if (parser == null)
        {
            parser = new ZsonParser(new Zson());
        }
        return parser;
    }
    public MemoBuilder setParser(IParser parser)
    {
        this.parser = parser;
        return this;
    }
    Encryption getEncryption()
    {
        if (encryption == null)
        {
            encryption = new ConcealEncryption("sfdgfgsgs");
            if (!encryption.init())
            {
                encryption = new NoEncryption();
            }
        }
        return encryption;
    }
    public MemoBuilder setEncryption(Encryption encryption)
    {
        this.encryption = encryption;
        return this;
    }
    ISerializer getSerializer()
    {
        if (serializer == null)
        {
            serializer = new MemoSerializer(getLogInterceptor());
        }
        return serializer;
    }
    public MemoBuilder setSerializer(ISerializer serializer)
    {
        this.serializer = serializer;
        return this;
    }
    public void build()
    {
        Memo.build(this);
    }

}

File: memo/src/main/java/com/zeoflow/memo/MemoBuilder.kt
package com.zeoflow.memo

import android.content.Context
import com.zeoflow.startup.ktx.ApplicationInitializer
import com.zeoflow.zson.Zson

class MemoBuilder @JvmOverloads constructor(context: Context = ApplicationInitializer.context) {
    private val context: Context

    init {
        MemoUtils.checkNull("Context", context)
        this.context = context.applicationContext
    }

    fun withLogInterceptor(logInterceptor: ILogInterceptor) : MemoBuilder {
        this.logInterceptor = logInterceptor
        return this
    }

    fun withSerializer(serializer: ISerializer) : MemoBuilder {
        this.serializer = serializer
        return this
    }

    fun withCryptoStorage(cryptoStorage: IStorage) : MemoBuilder {
        this.cryptoStorage = cryptoStorage
        return this
    }

    fun withParser(parser: IParser) : MemoBuilder {
        this.parser = parser
        return this
    }

    fun withConverter(converter: IConverter) : MemoBuilder {
        this.converter = converter
        return this
    }

    fun withEncryption(encryption: Encryption) : MemoBuilder {
        this.encryption = encryption
        return this
    }

    var logInterceptor: ILogInterceptor = object : ILogInterceptor {
        override fun onLog(message: String?) {

        }
    }
    var serializer: ISerializer = MemoSerializer(logInterceptor)
    var cryptoStorage: IStorage = SharedPreferencesStorage(context, STORAGE_TAG_DO_NOT_CHANGE)
    var parser: IParser = ZsonParser(Zson())
    var converter: IConverter = MemoConverter(parser)
    val storage: IStorage
        get() {
            return cryptoStorage
        }
    var encryption: Encryption = ConcealEncryption("default")
        get() {
            if (field is ConcealEncryption) {
                if (!(field as ConcealEncryption).init()) {
                    encryption = NoEncryption()
                }
            }
            return field
        }

    fun build() {
        Memo.build(this)
    }

    companion object {
        /**
         * NEVER ever change STORAGE_TAG_DO_NOT_CHANGE and TAG_INFO.
         * It will break backward compatibility in terms of keeping previous data
         */
        private const val STORAGE_TAG_DO_NOT_CHANGE = "MemoStorage"
    }
}